platform-tests/src/test/java/org/junit/platform/launcher/TagFilterTests.java

```diff
@@ -150,19 +150,49 @@ private void includeSingleTag(PostDiscoveryFilter filter) {
 		assertTrue(filter.apply(classWithTag1AndSurroundingWhitespace).included());
 		assertTrue(filter.apply(classWithBothTags).included());
 
+		assertThat(filter.apply(classWithTag1).getReason().get()).contains(
+			"Test included Because it contains tags: [tag1]");
+		assertThat(filter.apply(classWithTag1AndSurroundingWhitespace).getReason().get()).contains(
+			"Test included Because it contains tags: [tag1]");
+		assertThat(filter.apply(classWithBothTags).getReason().get()).contains(
+			"Test included Because it contains tags: [tag1,tag2]");
+
 		assertTrue(filter.apply(classWithTag2).excluded());
 		assertTrue(filter.apply(classWithDifferentTags).excluded());
 		assertTrue(filter.apply(classWithNoTags).excluded());
+
+		assertThat(filter.apply(classWithTag2).getReason().get()).contains(
+			"Test excluded Because it contains tags: [tag2]");
+		assertThat(filter.apply(classWithDifferentTags).getReason().get()).contains(
+			"Test excluded Because it contains tags: [foo,bar]");
+		assertThat(filter.apply(classWithNoTags).getReason().get()).contains(
+			"Test excluded Because it contains tags: []");
+
 	}
 
 	private void excludeSingleTag(PostDiscoveryFilter filter) {
 		assertTrue(filter.apply(classWithTag1).excluded());
 		assertTrue(filter.apply(classWithTag1AndSurroundingWhitespace).excluded());
 		assertTrue(filter.apply(classWithBothTags).excluded());
 
+		assertThat(filter.apply(classWithTag1).getReason().get()).contains(
+			"Test excluded Because it contains tags: [tag1]");
+		assertThat(filter.apply(classWithTag1AndSurroundingWhitespace).getReason().get()).contains(
+			"Test excluded Because it contains tags: [tag1]");
+		assertThat(filter.apply(classWithBothTags).getReason().get()).contains(
+			"Test excluded Because it contains tags: [tag1,tag2]");
```

Comment:
To make that a little bit more readable, you could maybe regroup both exclusion/reason assertions inside a method like `assertExcluded(FilterResult filterResult, String reason)` and call it here like `assertExcluded(filter.apply(...), "reason")`... (and have the `assertIncluded` counterpart for the inclusions)